### PR TITLE
SOS reports

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
The code can optionally use a "data:" URL for download, via `sos_use_data_url = true;`

This doesn't work too well, however.  Firefox freezes for a significant time between the click on the Download button and opening its own "Save" dialog.

**Comment by andreasn** [comment]
> consider adding more UI: name, case number, ticket number

I thought this was out of scope for now, but if you think it's crucial for the initial version, I'm happy to mock this up.


> consider adding support for profiles

What's profiles? I haven't come across this before.


> figure out whether/how to delete old reports

Does /var/tmp gets cleaned automatically somehow?
If not, we'll need some kind of purge button (or maybe it could do it automatically after a day or a week?).

**Comment by mvollmer** [comment]
>> consider adding more UI: name, case number, ticket number

> I thought this was out of scope for now, but if you think it's crucial for the initial version, I'm happy to mock this up.

I can't say how crucial this is, but it is also easy to do.  Just some text fields which get passed to sosreport as command line options.

**Comment by mvollmer** [comment]
>> consider adding support for profiles

> What's profiles? I haven't come across this before.

A profile determines what goes into the report.  Examples are "storage", "virt", "openstack", "desktop".
The idea would be that the support person on IRC or phone would tell you which profile to use.

This is not so easy to do as the name/case/ticket fields.

**Comment by mvollmer** [comment]
>> figure out whether/how to delete old reports

> Does /var/tmp gets cleaned automatically somehow?

It does actually get cleaned out, at least on Fedora.  See `/usr/lib/tmpfiles.d/tmp.conf` for the details.  Looks like stuff older than 30 days is removed.

However, we also have privacy concerns here.  Maybe people want to delete the reports as soon as possible.

**Comment by mvollmer** [comment]
Depends on #3120.

**Comment by mvollmer** [comment]
The created archive is now created in a 'secret' location and is always deleted when the dialog is closed.

**Comment by mvollmer** [comment]
Let's keep scraping in the PR, there is no API for progress and the final archive name.

One option is to add a "--status-fd" option to sosreport and write machine readable output to it.  We would redirect the real stdout to somewhere else and use "--status-fd 1".

**Comment by mvollmer** [comment]
I'll redo this with `systemd-run`...

**Comment by mvollmer** [comment]
> I'll redo this with `systemd-run`...

Nah, not now.  I was hoping that it would help with my problems getting sosreport killed dead, but it turns out that sosreport effectively ignores SIGTERM.

Using a systemd service for this has its own merits, however: Using a well-known one would allow us to survive a page reload and pick up a previous report generation.

**Comment by mvollmer** [comment]
State of the download machinery:

 - Files in /var/tmp/cockpit/download can be retrieved via GET to cockpit-ws when the login cookie is presented.  No other files are accessible.

 - The sosreport is put into a directory with a random name, which means the URL is unguessable.  The randomness of the name is currently not very good, though, and we need to improve it when we want to rely on this.

- The files in /var/tmp/cockpit/download can be retrieved via GET to cockpit-bridge, without authentication.

**Comment by stefwalter** [comment]
So thinking about this a bit more, I think we need something generic to do with channels and HTTP GET:

 1. Have a way to open a channel over the WebSocket but not multiplex the payload over the websocket.
 2. Pass an nonce style identifier back to the Cockpit javascrpt code.
 3. Use that nonce style identifier (ie: unpredictalbe) in the HTTP GET request to then receive the payload of that channel.

If this makes sense, we should probably use this mechanisms for the sideband Websocket channels we currently have.

**Comment by stefwalter** [comment]
@dperpeet @petervo CC on above ^^

**Comment by stefwalter** [comment]
Changed above to suggest POST to close out ]additional security cases. Although our CockpitWebServer doesn't support POST (any more) it may be worth adding it (back) for this case.

**Comment by stefwalter** [comment]
@mvollmer Do we have to link or redirect to the target URL in order to get sane file downloading behavior and browser UI? Or can one get that behavior from XMLHttpRequest? I assume not. If not, that closes out passing the nonce in headers.

**Comment by mvollmer** [comment]
This blocks on side-channel support for the downloads.

**Comment by mvollmer** [comment]
> Do we have to link or redirect to the target URL in order to get sane file downloading behavior and browser UI?  Or can one get that behavior from XMLHttpRequest?

I would expect the that the response to a XMLHttpRequest is always passed to JavaScript, regardless of Content-Type or Content-Disposition.  So we need a link or iframe to trigger the download.

The problem with a link is that the browser might navigate away from the current page, depending on the response.  One can use `target="_blank"` but causes some flicker.  One can also use the `download` attr.  I have to try that more.

**Comment by stefwalter** [comment]
Initial sideband work done in #3170 with another pull request to follow.

**Comment by stefwalter** [comment]
Further sideband work in #3196 with another pull request to follow.

**Comment by stefwalter** [comment]
Depends on #3199 and #3196 and #3201

**Comment by mvollmer** [comment]
This now uses #3201 for downloading.

**Comment by mvollmer** [comment]
The new fedora-23 failures are unrelated but need to be addressed.

```
Symbol not inside parenthesis at line 1 of /var/lib/selinux/targeted/tmp/modules/100/abrt/cil
Failed to parse /var/lib/selinux/targeted/tmp/modules/100/abrt/cil
libsemanage.semanage_load_files: Error while reading from file /var/lib/selinux/targeted/tmp/modules/100/abrt/cil. (No such file or directory).
FileNotFoundError: [Errno 2] No such file or directory
```

**Comment by mvollmer** [comment]
> The new fedora-23 failures are unrelated but need to be addressed.

This doesn't happen anymore with new fedora-23 images from a new virt-builder template.  Maybe the old template was causing this (doubtful), maybe something weird happened during creation of the old image (more likely).

**Comment by mvollmer** [comment]
Rebased and regenerated rhel-7 images.

File: pkg/sosreport/index.html [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+-->
+<html>
+  <head>
+    <title>Diagnostic reports</title>
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: Should be translatable.

File: pkg/sosreport/index.html [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    <meta charset="utf-8">
+    <link href="../base1/cockpit.css" rel="stylesheet">
+    <!-- <link href="sosreport.css" rel="stylesheet"> -->
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: Should we remove this?

File: pkg/sosreport/index.html [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+              <span class="alert-message" translatable="yes">
+                The generated archive contains data considered
+                sensitive and its content should be reviewed by the
+                originating organization before being passed to any
+                third party.
+              </span>
+            </div>
+            <div id="sos-progress">
+              <center>
+                <div translatable="yes" style="margin-bottom:5px">Generating report</div>
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: Should be translatable.

File: pkg/sosreport/index.html [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                <div class="progress">
+                  <div class="progress-bar" role="progressbar" style="width: 0%;"></div>
+                </div>
+              </center>
+            </div>
+            <div id="sos-download">
+              <center>
+                <div translatable="yes" style="margin-bottom:5px">Done!</div>
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: Are you sure we should have an exclam here? Was this in the design?

File: pkg/sosreport/index.html [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                <button class="btn btn-primary">Download report</button>
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: @andreasn Shouldn't this be a link rather than a button?

File: pkg/sosreport/index.html [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+              </center>
+            </div>
+            <div id="sos-error">
+              <div class="alert alert-danger">
+                <span class="pficon pficon-error-circle-o"></span>
+                <span class="alert-message"></span>
+              </div>
+              <textarea id="sos-error-extra" style="width:100%;max-width:100%;min-width:100%;min-height:200px">
```
Comment (stefwalter) [inline pkg/sosreport/index.html]: I would suggest against using inline styles here.

File: pkg/sosreport/index.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        $("#sos-progress .progress-bar").css("width", "0%");
+        $("#sos-download, #sos-error").hide();
+        $("#sos-alert, #sos-progress").show();
+        $("#sos-cancel").text(_("Cancel"));
+
+        sos_archive_url = null;
+        sos_archive_files = [ ];
+        sos_pid = null;
+
+        var task = cockpit.script('echo PID:$$; /usr/sbin/sosreport --batch', [ ],
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: This probably needs LANG=C if we're going to do screen scraping.

File: pkg/sosreport/index.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+                if (!sos_pid && (m = pid_regex.exec(output))) {
+                    sos_pid = m[1];
+                    output = output.slice(m[0].length);
+                }
+
+                p = 0;
+                while ((m = progress_regex.exec(output))) {
+                    p = (parseInt(m[1]) / parseInt(m[2])) * 100;
+                }
+                $("#sos-progress .progress-bar").css("width", p.toString()+"%");
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: Need spaces around the ```+```

File: pkg/sosreport/index.js [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                output += text;
+
+                if (!sos_pid && (m = pid_regex.exec(output))) {
+                    sos_pid = m[1];
+                    output = output.slice(m[0].length);
+                }
+
+                p = 0;
+                while ((m = progress_regex.exec(output))) {
+                    p = (parseInt(m[1]) / parseInt(m[2])) * 100;
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: parseInt needs a base of 10.

File: pkg/sosreport/index.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+                    sos_archive_files = [ archive, archive + ".md5" ];
+
+                    var query = window.btoa(JSON.stringify({
+                        payload: "fsread1",
+                        binary: "raw",
+                        path: archive,
+                        superuser: true,
+                        external: {
+                            "content-disposition": 'attachement; filename="' + basename + '"',
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: Does the basename need escaping or encoding? What can a valid Content-Disposition header file name include?

File: pkg/sosreport/index.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    function init() {
+        $(function () {
+            $("#sos").on("show.bs.modal", sos_init);
+            $("#sos-cancel").on("click", sos_cancel);
+            $('#sos-download button').on('click', sos_download);
+
+            // Establish a transport to the shell so that the tests
+            // know that we are ready to go.
+            //
+            cockpit.transport.wait(function () { });
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: The csrf_token above won't be valid until this succeeds. We should probably gate the entire init on this function, no?

File: pkg/sosreport/manifest.json [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{
+    "version": 0,
+
+    "tools": {
+        "index": {
+            "label": "Create diagnostic report"
```
Comment (stefwalter) [inline pkg/sosreport/manifest.json]: This needs to be marked as translatable somewhere. Currently I usually just use the same thing as the <title> and mark it as translatable.

File: pkg/sosreport/index.js [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        // We download via a hidden iframe to get better control over
+        // the error cases.
+        var iframe = $('<iframe>').attr('src', sos_archive_url);
+        iframe.on('load', function (event) {
+            var title = iframe.get(0).contentDocument.title;
+            if (title)
+                sos_error(title);
+        });
+        $('body').append(iframe);
+        iframe.hide();
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: Should we hide this before appending?

File: pkg/sosreport/index.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        });
+    }
+
+    function sos_cancel() {
+        if (sos_task) {
+            if (sos_pid) {
+                /* HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1278784
+                 * SIGTERM only aborts the current plugin, but SIGINT will stop the whole report
+                 */
+                cockpit.spawn([ "pkill", "-SIGINT", "-P", sos_pid ], { superuser: true, err: "message" }).
```
Comment (stefwalter) [inline pkg/sosreport/index.js]: Why do we use pkill here? This means yet another runtime dependency. A stream channel will kill its process if it receives close with a problem. Should we make that more explicit?

File: pkg/sosreport/index.html (hunk 5 above)
Comment (andreasn) [inline pkg/sosreport/index.html]: Do you mean since it's a link?
That could work too, but wouldn't it de-empathise the link too much?

Comment (stefwalter) [inline pkg/sosreport/index.html]: Yes, that's true.

File: pkg/sosreport/index.html (hunk 1 above)
Comment (mvollmer) [inline pkg/sosreport/index.html]: Done.

File: pkg/sosreport/index.html (hunk 3 above)
Comment (mvollmer) [inline pkg/sosreport/index.html]: It is, isn't it?

File: pkg/sosreport/index.html (hunk 4 above)
Comment (mvollmer) [inline pkg/sosreport/index.html]: Yes: https://trello-attachments.s3.amazonaws.com/55926ff75f627855f4f69410/1573x2097/b09aafbdd876af5fdc4393934c033876/create-sos-report-v2.png

Comment (mvollmer) [inline pkg/sosreport/index.html]: @andreasn, what do you think?

File: pkg/sosreport/index.html (hunk 6 above)
Comment (mvollmer) [inline pkg/sosreport/index.html]: Done,

File: pkg/sosreport/index.html (hunk 2 above)
Comment (mvollmer) [inline pkg/sosreport/index.html]: It's now used for the #sos-error-extra styles.

File: pkg/sosreport/index.js (hunk 7 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: Good point, done.

File: pkg/sosreport/index.js (hunk 8 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: Done.

File: pkg/sosreport/index.js (hunk 9 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: Done.

File: pkg/sosreport/index.js (hunk 13 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: Yes, done.

File: pkg/sosreport/index.js (hunk 11 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: Yes, done.

**Comment by mvollmer** [comment]
Depends on #3272.

File: pkg/sosreport/index.js (hunk 11 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: See #3272.

File: pkg/sosreport/index.js (hunk 14 above)
Comment (mvollmer) [inline pkg/sosreport/index.js]: You are right, `task.close("cancelled")` is enough to kill sosreport, although https://bugzilla.redhat.com/show_bug.cgi?id=1278784 still exists.  I can't really explain this...

File: pkg/sosreport/manifest.json (hunk 12 above)
Comment (mvollmer) [inline pkg/sosreport/manifest.json]: I have used it as the dialog title.

**Comment by mvollmer** [comment]
> Depends on #3272.

That doesn't work so well, so let's use `cockpit.transport.wait`.

**Pull request closed without merging**